File: packages/typescript/src/core/validation.ts
import { z } from 'zod';
import { ValidationError } from './errors';
import type { FingerprintSubmission } from './types';

/** Hex string pattern matching proto's hex validation */
const hexPattern = /^[0-9a-fA-F]+$/;

/** UUID v4 pattern (loose — accepts any UUID-shaped string) */
const uuidPattern =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

/** Zod schema for FingerprintValue, matching proto validation rules */
export const FingerprintValueSchema = z.object({
  orgId: z.string().regex(uuidPattern, 'orgId must be a valid UUID'),
  tenantId: z.string().regex(uuidPattern, 'tenantId must be a valid UUID'),
  eventId: z.string().regex(uuidPattern, 'eventId must be a valid UUID'),
  signerId: z
    .string()
    .min(64, 'signerId must be at least 64 characters')
    .max(140, 'signerId must be at most 140 characters')
    .regex(hexPattern, 'signerId must be hex-encoded')
    .optional(),
  documentId: z
    .string()
    .min(1, 'documentId is required')
    .max(256, 'documentId must be at most 256 characters'),
  documentRef: z
    .string()
    .min(32, 'documentRef must be at least 32 characters')
    .max(128, 'documentRef must be at most 128 characters')
    .regex(hexPattern, 'documentRef must be hex-encoded'),
  timestamp: z.string().min(1, 'timestamp is required'),
  version: z.number().int().min(1, 'version must be >= 1'),
});

/** Zod schema for DedSignatureProof, matching proto SignatureProof */
export const SignatureProofSchema = z.object({
  id: z
    .string()
    .min(1, 'id is required')
    .max(140, 'id must be at most 140 characters')
    .regex(hexPattern, 'id must be hex-encoded'),
  signature: z
    .string()
    .min(64, 'signature must be at least 64 characters')
    .max(2048, 'signature must be at most 2048 characters')
    .regex(hexPattern, 'signature must be hex-encoded'),
  algorithm: z.literal('SECP256K1_RFC8785_V1'),
});

/** Zod schema for FingerprintMetadata */
export const FingerprintMetadataSchema = z.object({
  hash: z
    .string()
    .min(32)
    .max(128)
    .regex(hexPattern, 'hash must be hex-encoded'),
  tags: z
    .record(
      z.string().min(1).max(32),
      z.string().max(32)
    )
    .refine((tags) => Object.keys(tags).length <= 6, {
      message: 'tags must have at most 6 pairs',
    })
    .optional(),
});

/** Zod schema for a complete FingerprintSubmission */
export const FingerprintSubmissionSchema = z.object({
  attestation: z.object({
    content: FingerprintValueSchema,
    proofs: z
      .array(SignatureProofSchema)
      .min(1, 'at least one proof is required'),
  }),
  metadata: FingerprintMetadataSchema.optional(),
});

/**
 * Validate a FingerprintSubmission, throwing a ValidationError if invalid.
 *
 * @param submission - The submission to validate
 * @throws ValidationError with structured issue list
 */
export function validateSubmission(submission: FingerprintSubmission): void {
  const result = FingerprintSubmissionSchema.safeParse(submission);
  if (!result.success) {
    const issues = result.error.issues.map((issue) => ({
      path: issue.path.join('.'),
      message: issue.message,
    }));
    throw new ValidationError('Invalid FingerprintSubmission', issues);
  }
}

/**
 * Validate a FingerprintSubmission without throwing.
 *
 * @param submission - The submission to validate
 * @returns Object with `success` boolean and optional `issues` array
 */
export function safeValidateSubmission(submission: FingerprintSubmission): {
  success: boolean;
  issues?: Array<{ path: string; message: string }>;
} {
  const result = FingerprintSubmissionSchema.safeParse(submission);
  if (result.success) {
    return { success: true };
  }
  return {
    success: false,
    issues: result.error.issues.map((issue) => ({
      path: issue.path.join('.'),
      message: issue.message,
    })),
  };
}

File: packages/python/src/constellation_digital_evidence_sdk/core/validation.py
"""Validation helpers matching protobuf constraints."""

from __future__ import annotations

import re
from typing import Any

from .errors import ValidationError
from .types import FingerprintSubmission

_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")
_UUID_RE = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}"
    r"-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)


def _check(
    condition: bool, path: str, message: str, issues: list[dict[str, str]]
) -> None:
    if not condition:
        issues.append({"path": path, "message": message})


def validate_submission(submission: FingerprintSubmission) -> None:
    """Validate a FingerprintSubmission, raising ValidationError if invalid.

    Args:
        submission: The submission to validate.

    Raises:
        ValidationError: With structured issue list if validation fails.
    """
    issues: list[dict[str, str]] = []
    att = submission.attestation
    v = att.content

    # FingerprintValue validation
    _check(bool(_UUID_RE.match(v.org_id)), "orgId", "must be a valid UUID", issues)
    _check(bool(_UUID_RE.match(v.tenant_id)), "tenantId", "must be a valid UUID", issues)
    _check(bool(_UUID_RE.match(v.event_id)), "eventId", "must be a valid UUID", issues)
    _check(len(v.document_id) >= 1, "documentId", "is required", issues)
    _check(len(v.document_id) <= 256, "documentId", "must be at most 256 chars", issues)
    _check(
        32 <= len(v.document_ref) <= 128,
        "documentRef",
        "must be 32-128 characters",
        issues,
    )
    _check(
        bool(_HEX_RE.match(v.document_ref)),
        "documentRef",
        "must be hex-encoded",
        issues,
    )
    _check(bool(v.timestamp), "timestamp", "is required", issues)
    _check(v.version >= 1, "version", "must be >= 1", issues)

    if v.signer_id is not None:
        _check(
            64 <= len(v.signer_id) <= 140,
            "signerId",
            "must be 64-140 characters",
            issues,
        )
        _check(
            bool(_HEX_RE.match(v.signer_id)),
            "signerId",
            "must be hex-encoded",
            issues,
        )

    # Proofs validation
    _check(len(att.proofs) >= 1, "proofs", "at least one proof is required", issues)
    for i, proof in enumerate(att.proofs):
        prefix = f"proofs[{i}]"
        _check(
            1 <= len(proof.id) <= 140,
            f"{prefix}.id",
            "must be 1-140 characters",
            issues,
        )
        _check(
            bool(_HEX_RE.match(proof.id)),
            f"{prefix}.id",
            "must be hex-encoded",
            issues,
        )
        _check(
            64 <= len(proof.signature) <= 2048,
            f"{prefix}.signature",
            "must be 64-2048 characters",
            issues,
        )
        _check(
            proof.algorithm == "SECP256K1_RFC8785_V1",
            f"{prefix}.algorithm",
            "must be SECP256K1_RFC8785_V1",
            issues,
        )

    # Metadata validation (if present)
    meta = submission.metadata
    if meta is not None:
        _check(
            32 <= len(meta.hash) <= 128,
            "metadata.hash",
            "must be 32-128 characters",
            issues,
        )
        _check(
            bool(_HEX_RE.match(meta.hash)),
            "metadata.hash",
            "must be hex-encoded",
            issues,
        )
        if meta.tags is not None:
            _check(
                len(meta.tags) <= 6,
                "metadata.tags",
                "must have at most 6 pairs",
                issues,
            )
            for key, val in meta.tags.items():
                _check(
                    1 <= len(key) <= 32,
                    f"metadata.tags.{key}",
                    "key must be 1-32 characters",
                    issues,
                )
                _check(
                    len(val) <= 32,
                    f"metadata.tags.{key}",
                    "value must be at most 32 characters",
                    issues,
                )

    if issues:
        raise ValidationError("Invalid FingerprintSubmission", issues)
